---
title: "manhattan plots"
author: "Jason Torres"
date: "January 29, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)

library("data.table")
library("tidyverse")
library("grid")
library("gridExtra")
library("ggplot2")
library("cowplot")
library("ggrepel")


local.dir <- "/Users/jtorres/Google Drive/Science/Projects/metaxcan_t2d_v2/"
rescomp.dir <- "/Users/jtorres/FUSE5/projects/metaxcan_t2d_v2/"
df.dir <- rescomp.dir %&% "data_frames/"
assoc.df <- fread(df.dir %&% "association_results.txt")
islet.df <- fread(df.dir %&% "islet_association.txt")

loci.df <- fread(df.dir %&% "gwas_windows.txt")

gwas.names <- c("scott","scottadj","fi","fg")
tiss.names <- c("panc","liv","mus","adi","multi")
tiss.df <- data.frame(abrev=tiss.names,
                      full.name=c("Pancreas","Liver",
                                  "Muscle-skeletal","Adipose-subcutaneous",
                                  "cross-tissue"))

plot.dir <- rescomp.dir %&% "plots/"

```




```{r}

manhat <- function(dataframe, title=NULL, max.y="max", suggestiveline=0, genomewideline=-log10(5e-8), size.x.labels=9, size.y.labels=10, annotate=F, SNPlist=NULL, rev=FALSE,col1,col2,trunc.y=NA,START=1,CAP=0,BY=1) {
  #BY is the interval for the Y-axis 
  
	#if (annotate & is.null(SNPlist)) stop("You requested annotation but provided no SNPlist!")
	d=dataframe
	
	#limit to only chrs 1-23?
	d=d[d$CHR %in% 1:23, ]
	#d=d[sample(1:dim(d)[1],5000,replace = FALSE),] # To speed up testing ONLY !!! 
  d <- arrange(d,CHR,BP) 
	
	if ("CHR" %in% names(d) & "BP" %in% names(d) & "P" %in% names(d) ) {
	  d <- dplyr::select(d,one_of("CHR","BP","P"))
		d=na.omit(d)
		d=d[d$P>0 & d$P<=1, ]
		d$logp = -log10(d$P)
		
		d$pos=NA
		ticks=NULL
		lastbase=0
		
		#new 2010-05-10
		numchroms=length(unique(d$CHR))
		if (numchroms==1) {
			d$pos=d$BP
		} else {
		
			for (i in unique(d$CHR)) {
				if (i==1) {
					d[d$CHR==i, ]$pos=d[d$CHR==i, ]$BP
				}	else {
					lastbase=lastbase+tail(subset(d,CHR==i-1)$BP, 1)
					d[d$CHR==i, ]$pos=d[d$CHR==i, ]$BP+lastbase
				}
				ticks=c(ticks, d[d$CHR==i, ]$pos[floor(length(d[d$CHR==i, ]$pos)/2)+1])
			}
			ticklim=c(min(d$pos),max(d$pos))

		}
		
		mycols=rep(c("gray10","gray60"),max(d$CHR))
		
		if (max.y=="max") maxy=ceiling(max(d$logp)) else maxy=max.y
		if (maxy<8) maxy=8
		if (!is.na(trunc.y)) maxy=trunc.y

#		if (annotate) d.annotate=d[as.numeric(substr(d$SNP,3,100)) %in% SNPlist, ]
		
		if (numchroms==1) {
			plot=qplot(pos,logp,data=d,ylab=expression(-log[10](italic(p))), xlab=paste("Chromosome",unique(d$CHR),"position"))
		}	else {
			plot <- ggplot(data=d,aes(x=pos,y=logp,color=factor(CHR),fill=factor(CHR))) + 
			  geom_point(shape=21) + ylab(expression(-log[10](italic(p)))) + 
			  scale_x_continuous(name="Chromosome", breaks=ticks, labels=(unique(d$CHR))) + 
			  scale_y_continuous(limits=c(0,maxy), breaks=seq(START,maxy+CAP,by=BY), labels=seq(START,maxy+CAP,by=BY)) +
			  #scale_y_continuous(limits=c(0,MAX), breaks=seq(START,MAX,by=BY), labels=seq(START,MAX,by=BY)) +
			  theme(legend.position = "none") + theme(title=title) +
			  scale_fill_manual(values = rep(c(col1,col2),11)) + 
			  scale_color_manual(values = rep(c(col1,col2),11)) + 
			  theme(panel.background=element_blank(), 
			         panel.grid.minor=element_blank(),
			         axis.text.x=element_text(size=size.x.labels, colour="black"), 
			         axis.text.y=element_text(size=size.y.labels, colour="black"),
			         panel.border = element_rect(colour = "black", fill=NA, size=2))
			if (rev==TRUE){
			  plot <- plot + scale_y_reverse(breaks=seq(START,maxy+CAP,by=BY),labels=seq(START,maxy+CAP,by=BY),limits=c(maxy,0)) 
			}
		}
		if (suggestiveline) plot=plot+geom_hline(yintercept=suggestiveline,colour="gray", alpha=I(1/3))
		if (genomewideline) plot=plot+geom_hline(yintercept=genomewideline,colour="red")
		plot
	}	else {
		stop("Make sure your data frame contains columns CHR, BP, and P")
	}
  return(list(plot,d))
}


```





Bonferroni Significant data frames 

```{r}

get_manhat <- function(gwas.name,tiss.abrev,pred.perf.q=0.05,
                       col1,col2,cap=0,by=1){
  sub.df <- dplyr::select(assoc.df,one_of("gene","gene_name","chrom",
                                          "gene.start","gene.end"),
                          matches(tiss.abrev%&%"."%&%gwas.name%&%"\\.")) %>%
    as.data.frame(.)

  if (tiss.abrev!="multi"){
    # Filter out NA's and genes with ppq > 0.05
    index <- which(names(sub.df)==tiss.abrev%&%"."%&%gwas.name%&%".qpp") 
    keep.vec <- sub.df[,index] <= pred.perf.q
    keep.vec[is.na(keep.vec)] <- FALSE
    sub.df <- sub.df[keep.vec,]    
  }
  bonfer <- 0.05 / dim(sub.df)[1]
  # keep bonferonni associations 
  index <- which(names(sub.df)==tiss.abrev%&%"."%&%gwas.name%&%".p") 
  keep.vec <- sub.df[,index] <= 1
  keep.vec[is.na(keep.vec)] <- FALSE
  sub.df <- sub.df[keep.vec,]
  if (tiss.abrev!="multi"){
    len <- dim(sub.df)[2]
    names(sub.df)[(len-4):len] <- c("z","P","pred.perf.q","p_H3","p_H4")
  } else{
    sub.df$pred.perf.q <- NA; sub.df$p_H3 <- NA; sub.df$p_H4<-NA
    len <- dim(sub.df)[2]
    names(sub.df)[(len-4):(len-3)] <- c("z","P")
  }
  if (dim(sub.df)[1]>0){
    sub.df$tissue <- filter(tiss.df,abrev==tiss.abrev)$full.name[1]
    sub.df$tissue <- as.character(sub.df$tissue)    
  }
  names(sub.df)[names(sub.df)=="chrom"] <- "CHR"
  names(sub.df)[names(sub.df)=="gene.start"] <- "BP"
  
  plot.df <- sub.df %>% dplyr::select(one_of("CHR","BP","P"))
  plt.list <- manhat(dataframe = sub.df,col1=col1,col2=col2,
         genomewideline=-log10(bonfer),CAP=cap,BY=by)
  df <- plt.list[[2]]
  pb <- txtProgressBar(min=0,max=dim(df)[1],style=3)
  df$gene_name <- map(1:dim(df)[1],function(i){
    setTxtProgressBar(pb,i)
    row.df <- df[i,]
    filter(sub.df,CHR==row.df$CHR,BP==row.df$BP)$gene_name
  }) %>% as.character(.)
  close(pb)
  plt.list[[1]] + geom_text_repel(data=filter(df,P<=bonfer),
                     aes(x=pos,y=logp,label=gene_name),size=2)
}

get_manhat_islet <- function(gwas.name,col1,col2,cap=0,by=1){
  sub.df <- dplyr::select(islet.df,one_of("gene","gene_name","chrom",
                                          "gene.start","gene.end"),
                          matches(gwas.name%&%"\\.")) %>%
    as.data.frame(.)

  bonfer <- 0.05 / dim(sub.df)[1]
  # keep bonferonni associations 
  index <- which(names(sub.df)==gwas.name%&%".p") 
  keep.vec <- sub.df[,index] <= 1
  keep.vec[is.na(keep.vec)] <- FALSE
  sub.df <- sub.df[keep.vec,]

  names(sub.df)[names(sub.df)=="chrom"] <- "CHR"
  names(sub.df)[names(sub.df)=="gene.start"] <- "BP"
  names(sub.df)[7] <- "P"
  plot.df <- sub.df %>% dplyr::select(one_of("CHR","BP","P"))
  plt.list <- manhat(dataframe = sub.df,col1=col1,col2=col2,
         genomewideline=-log10(bonfer),CAP=cap,BY=by)
  df <- plt.list[[2]]
  pb<-txtProgressBar(min=0,max=dim(df)[1],style=3)
  df$gene_name <- map(1:dim(df)[1],function(i){
    setTxtProgressBar(pb,i)
    row.df <- df[i,]
    filter(sub.df,CHR==row.df$CHR,BP==row.df$BP)$gene_name
  }) %>% as.character(.)
  close(pb)
  
  plt.list[[1]] + geom_text_repel(data=filter(df,P<=bonfer),
                     aes(x=pos,y=logp,label=gene_name),size=2)
}

save_manhat_plots <- function(gwas.name){
  plt.multi <- get_manhat(gwas.name,"multi",col1="dodgerblue4",col2="dodgerblue1",by=2)
  plt.isl <- get_manhat_islet(gwas.name,col1="slategray4",col2="slategray3")
  plt.panc <- get_manhat(gwas.name,"panc",col1="chartreuse4",col2="chartreuse3")
  plt.mus <- get_manhat(gwas.name,"mus",col1="coral3",col2="coral4")
  plt.liv <- get_manhat(gwas.name,"liv",col1="rosybrown4",col2="rosybrown3")
  plt.adi <- get_manhat(gwas.name,"adi",col1="goldenrod1",col2="goldenrod3")
  plt <- arrangeGrob(plt.adi,plt.mus,plt.liv,plt.panc,plt.multi,ncol=1)
  plt2 <- arrangeGrob(plt.adi,plt.mus,plt.liv,plt.panc,plt.isl,plt.multi,ncol=1)
  ggplot2::ggsave(filename=plot.dir%&%gwas.name%&%".manhattan-tracks.png",plt,
        width=14,height=15)
  ggplot2::ggsave(filename=plot.dir%&%gwas.name%&%".manhattan-tracks_full.png",plt2,
        width=14,height=15)
}


```


```{r}

save_manhat_plots(gwas.name="scott")
save_manhat_plots(gwas.name="scottadj")

```



