---
title: "Untitled"
author: "Jason Torres"
date: "February 8, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Setup 


```{r}

"%&%" <- function(a,b) paste0(a,b)

fuse.path <- "/home/jason/science/servers/"
#fuse.path <- "/Users/jtorres/"
rescomp.dir <- fuse.path %&% "FUSE5/projects/metaxcan_t2d_v2/"
df.dir <- rescomp.dir %&% "data_frames/"
plot.dir <- rescomp.dir %&% "plots/"
ref.dir <- rescomp.dir %&% "reference_files/"

library("data.table")
assoc.df <- fread(df.dir %&% "association_results.txt")
islet.df <- fread(df.dir %&% "islet_association.txt")

lws.scott.df <- fread(df.dir %&% "Scott_T2D_lws-significant.full.txt")
loc.prof.df <- fread(df.dir %&% "locus-profile.txt")
sigloc.df <- fread(df.dir %&% "significant-loci.txt")
library("tidyverse")
known.genes <- filter(loc.prof.df,reported_gene==TRUE)$gene_name %>% unique(.)

```


```{r}

gwas.df <- fread(df.dir %&% "DIAGRAM_T2D_SCOTT.ma")
names(gwas.df)[1] <- "RSID"
snps.df <- fread("cat " %&% df.dir %&% "snp_keyfile.txt.gz" %&%  " | zmore")
gwas.full.df <- inner_join(gwas.df,snps.df,by="RSID")

```



```{r}

library("rtracklayer")
gtf.gr <- rtracklayer::import(ref.dir%&%"gencode.v19.genes.v7.patched_contigs.gtf")
gtf.df <- as.data.frame(gtf.gr,stringsAsFactors=FALSE)
```




```{r}

library("ggbio")

gene_annot_plot <- function(chromo,loc.start,loc.end,sub.df){
  g.df <- filter(gtf.df,seqnames==chromo,
                 start<=loc.end,end>=loc.start,type=="gene",gene_type=="protein_coding")
  g.df$y <- sample(1:dim(g.df)[1],size=dim(g.df)[1],replace=FALSE)
  g.df$direction <- ifelse(g.df$strand=="+","first",
                           ifelse(g.df$strand=="-","last",NA))
  g.df$color <- ifelse(g.df$gene_name%in%known.genes,"firebrick3",
                       ifelse(g.df$gene_name%in%sub.df$gene_name,"black","grey"))
  plt <- ggplot(data=g.df) + geom_segment(aes(x=start,y=y,xend=end,yend=y),
                                   arrow=arrow(ends = g.df$direction,
                                               length = unit(0.03, "npc")),color=g.df$color) + 
    theme_bw() + theme(panel.grid.minor = element_blank(),
                       panel.grid.major = element_blank(),
                       axis.text.y=element_blank(),
                       axis.ticks.y=element_blank()) + ylab("") + 
    geom_text(aes(x=start,y=y,label=gene_name),fontface="italic",size=1.8,nudge_y=0.5,
              color=g.df$color)
  return(plt)
}


```


```{r}

gwas_loc_plot <- function(chromo,loc.start,loc.end){
 sub.df <- filter(gwas.full.df,CHR=="chr"%&%chromo,POS>=loc.start,POS<=loc.end)
 bonfer <- 5*10^(-8)
 sub.df$col <- ifelse(sub.df$p<=bonfer,"firebrick3","azure4")
 sub.df$fil <- ifelse(sub.df$p<=bonfer,"firebrick1","azure3")
 plt <- ggplot(data=sub.df,aes(x=POS,-log(p,base=10))) + 
   geom_point(shape=21,fill=sub.df$fil,color=sub.df$col) + 
   theme_clear() + 
   geom_hline(yintercept = -log(bonfer,base=10),linetype=2,color="dimgrey") + 
    ylab(expression(paste("-log"[10],"(p)"))) 
 return(plt)
}


```



```{r}

library("viridis")
library("ggrepel")
library("gridExtra")
library("rlist")

loc_plot <- function(input.df,tiss.abrev,top){
  vec <- input.df$p
  col.vec <- ifelse((input.df$z %>% 
                       sign(.))>0,viridis(2)[2],viridis(2)[1])
  bonfer <- 0.05/(vec %>% na.omit(.) %>% length(.))
  input.df$mid.point <- (input.df$gene.end+input.df$gene.start)/2
  input.df$nlp <- -log(input.df$p,base=10)
  sig.df <- filter(input.df,p<=bonfer)
  plt <- ggplot(data=input.df,aes(x=mid.point,y=nlp)) + 
  geom_point(size=1.5,shape=21,fill=col.vec) + 
  scale_shape_manual(values=c(21,23)) + 
  geom_hline(yintercept=-log(bonfer,base=10),
            color="black",size=0.3,linetype=2) + 
  theme_bw() + theme(legend.position = "none",
                    panel.grid.major = element_blank(),
                   panel.grid.minor = element_blank(),
                     axis.title.y=element_text(size=5),
                      axis.text.y=element_text(size=3)) + 
  xlab("position") + 
  ylab(expression(paste("-log"[10],"(p)"))) 

  if (dim(sig.df)[1]>0){
    sig.cols <- ifelse(sig.df$gene_name%in%known.genes,"grey99","grey88")
    plt <- plt + geom_label_repel(data=sig.df,aes(x=mid.point,
    y=-log(p,base=10),
    label=gene_name),fontface="italic",
    label.padding=0.1,size=1.5,fill=sig.cols) 
  }
    
  if (top>25){
    plt <- plt + scale_y_continuous(breaks=seq(0,top,by=10),
                                    limits=c(0,top))    
  } else if(top>10&top<25){
    plt <- plt + scale_y_continuous(breaks=seq(0,top,by=5),
                                    limits=c(0,top))
  } else{
    plt <- plt + scale_y_continuous(breaks=seq(0,top,by=2),
                                    limits=c(0,top))    
  }

  titl<-ifelse(tiss.abrev=="multi","TissueXcan (GTEx)",
               ifelse(tiss.abrev=="panc","Pancreas",
                      ifelse(tiss.abrev=="liv","Liver",
                             ifelse(tiss.abrev=="adi",
                                    "Adipose-Subcutaneous",
                                    ifelse(tiss.abrev=="isl",
                                           "Islets",
                                           "Muscle-Skeletal")))))
  xpoint <- min(input.df$gene.start)+((max(input.df$gene.end)-min(input.df$gene.start))/8)
  ypoint <- top * 9.8/10 
  plt <- plt + annotate("text",x=xpoint,y=ypoint,
                        label=titl,size=2.5)
  evalu <- dim(sig.df)[1] > 0
  return(list(evalu,plt))
}

locus_track_plot <- function(loc.id){
  l.df <- filter(loc.prof.df,locus.id==loc.id)
  chromo <- l.df$chrom %>% unique(.)
  loc.start <- l.df$loc.start %>% unique(.)
  loc.end <- l.df$loc.end %>% unique(.)
  sub.df <- filter(assoc.df,chrom==chromo,gene.end >= loc.start, gene.start<=loc.end)
  isl.df <- filter(islet.df,chrom==chromo,gene.end>=loc.start,
                   gene.start<=loc.end)
  sub.df$reported_gene <- (sub.df$gene_name %in% known.genes)
  isl.df$reported_gene <- (isl.df$gene_name %in% known.genes)

  top <- min(c(min(na.omit(sub.df$multi.scott.p)),
               min(na.omit(sub.df$panc.scott.p)),
      min(na.omit(sub.df$liv.scott.p)),
      min(na.omit(sub.df$mus.scott.p)),
      min(na.omit(sub.df$adi.scott.p)),
      min(na.omit(isl.df$scott.p)))) %>% -log(.,base=10) %>% ceiling(.)
  top <- (top + (top*0.10)) %>% ceiling(.) 
  plt.gwas <- gwas_loc_plot(chromo,loc.start,loc.end)
  # Multi 
  in.df <- dplyr::select(sub.df,
                         one_of("gene","gene_name","chrom","gene.start",
                                "gene.end","reported_gene",
                                "multi.scott.p","multi.scott.z"))
  names(in.df) <- c("gene","gene_name","chrom","gene.start",
                    "gene.end","reported_gene","p","z")
  plt.multi <- loc_plot(in.df,"multi",top)
  # Pancreas 
  in.df <- dplyr::select(sub.df,
                         one_of("gene","gene_name","chrom","gene.start",
                                "gene.end","reported_gene",
                                "panc.scott.p","panc.scott.z"))
  names(in.df) <- c("gene","gene_name","chrom","gene.start",
                    "gene.end","reported_gene","p","z")
  plt.panc <- loc_plot(in.df,"panc",top)
  # Liver 
  in.df <- dplyr::select(sub.df,
                         one_of("gene","gene_name","chrom","gene.start",
                                "gene.end","reported_gene",
                                "liv.scott.p","liv.scott.z"))
  names(in.df) <- c("gene","gene_name","chrom","gene.start",
                    "gene.end","reported_gene","p","z")
  plt.liv <- loc_plot(in.df,"liv",top)
  # Muscle 
  in.df <- dplyr::select(sub.df,
                         one_of("gene","gene_name","chrom","gene.start",
                                "gene.end","reported_gene",
                                "mus.scott.p","mus.scott.z"))
  names(in.df) <- c("gene","gene_name","chrom","gene.start",
                    "gene.end","reported_gene","p","z")
  plt.mus <- loc_plot(in.df,"mus",top)
  # Adipose 
  in.df <- dplyr::select(sub.df,
                         one_of("gene","gene_name","chrom","gene.start",
                                "gene.end","reported_gene",
                                "adi.scott.p","adi.scott.z"))
  names(in.df) <- c("gene","gene_name","chrom","gene.start",
                    "gene.end","reported_gene","p","z")
  plt.adi <- loc_plot(in.df,"adi",top)
  # Islet 
  in.df <- dplyr::select(isl.df,
                         one_of("gene","gene_name","chrom","gene.start",
                                "gene.end","reported_gene",
                                "scott.p","scott.z"))
  names(in.df) <- c("gene","gene_name","chrom","gene.start",
                    "gene.end","reported_gene","p","z")
  plt.isl <- loc_plot(in.df,"isl",top)
  detach("package:ggbio", unload=TRUE)
  gene.plt <- gene_annot_plot(chromo,loc.start,loc.end,sub.df)
  library("ggbio")

  name.vec <- c("plt.multi","plt.isl","plt.panc","plt.mus","plt.adi","plt.liv")
  eval.vec <- c(plt.multi[[1]], plt.isl[[1]],plt.panc[[1]], plt.mus[[1]],
                plt.adi[[1]],plt.liv[[1]])
  keep.plots <- name.vec[eval.vec]
  
  p.list <- map(keep.plots,function(p){
    eval(parse(text=p%&%"[[2]]"))
  })
  h <- 1/length(p.list)
  height.vec <- c(rep(h,length(p.list)),1.5*h,h)
  plt <- tracks(list.append(p.list,plt.gwas,gene.plt),
                heights=height.vec,main="Region: "%&%loc.id) + 
    scale_x_sequnit("Mb") + 
    theme(axis.text.x=element_text(size=5),
          axis.text.y=element_text(size=5),
          axis.title.y=element_text(size=8))  

  ggsave(plot=plt,
         filename=plot.dir%&%"Region_"%&%loc.id%&%".png",
         width=5,height=6)

}


save_loc_track_plots <- function(remove.vec=c(8,25,28)){
  loc.ids <- loc.prof.df$locus.id %>% unique(.)
  loc.ids <- loc.ids[!(loc.ids%in%remove.vec)]
  pb <- txtProgressBar(min=0,max=length(loc.ids),style=3)
  for (i in 1:length(loc.ids)){
    locid <- loc.ids[i]
    print(locid)
    setTxtProgressBar(pb,i)
    locus_track_plot(locid)
  }

}


```


```{r}

save_loc_track_plots()

```



```{bash}

#convert loc-TissueXcan_only-known_*.pdf loc-TissueXcan_only-known.pdf
#convert loc-TissueXcan_only-novel_*.pdf loc-TissueXcan_only-novel.pdf
#convert loc-TissueXcan_shared_*.pdf loc-TissueXcan_shared.pdf

```

