---
title: "manuscript_results"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)

library("data.table")
library("tidyverse")

fuse.path <- "/home/jason/science/servers/"
#fuse.path <- "/Users/jtorres/"
server.dir <- fuse.path %&% "FUSE5/" 
work.dir <- server.dir %&% "projects/metaxcan_t2d_v2/"
dat.dir <- work.dir %&% "data_frames/"

```


```{r}

assoc.df <- fread(dat.dir %&% "association_results.txt")
islet.df <- fread(dat.dir %&% "islet_association.txt")

gws.df <- fread(dat.dir %&% "result-table_ScottT2D_gws.txt")
lws.df <- fread(dat.dir %&% "result-table_ScottT2D_lws.txt")

```



# Tissue-wide association analyses - Genome-wide significance 

## Significant associations and genes 

### Individual GTEx tissues (adipose, muscle, liver, pancreas)

```{r}

# Number of test performed in each tissue 
assoc.df$adi.scott.z %>% as.numeric(.) %>% na.omit(.) %>% length(.) # adipose: 7243
assoc.df$liv.scott.z %>% as.numeric(.) %>% na.omit(.) %>% length(.) # liver: 2502
assoc.df$mus.scott.z %>% as.numeric(.) %>% na.omit(.) %>% length(.) # muscle: 6509
assoc.df$panc.scott.z %>% as.numeric(.) %>% na.omit(.) %>% length(.) # pancreas: 4603
assoc.df$multi.scott.z %>% as.numeric(.) %>% na.omit(.) %>% length(.) # TissueXcan: 17433
islet.df$scott.z %>% as.numeric(.) %>% na.omit(.) %>% length(.) # Islets: 4322

# Total number of significant associations across all GTEx tissues of interest (adipose, muscle, liver, pancreas) 

(filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets"))) %>% dim(.))[1] # 28

# Number of associatiosn per tissue 

(filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets")))$tissue %>% table(.)) # 45 

# Number of unique genes 

filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets")))$gene_name %>% unique(.) %>% length(.) # 22

gene.tab <- filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets")))$gene_name %>% table(.) %>% sort(.) %>% rev(.) 
# Genes that show up in multiple tissues 
count <- sum(gene.tab>1)
filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets")),gene_name%in%names(gene.tab[1:count])) # looking up tissues 

# Number of genes that have been previously reported 

filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets")),gene.type=="reported")$gene_name %>% unique(.) %>% length(.)
filter(gws.df,!(tissue%in%c("TissueXcan (GTEx)","Islets")),gene.type=="reported")$gene_name %>% unique(.)

```


### TissueXcan to leverage information content across 44 GTEx tissues 

```{r}

# Number of sig associations 
(filter(gws.df,tissue == "TissueXcan (GTEx)") %>% dim(.))[1] # 29 

txn.genes <- filter(gws.df,tissue == "TissueXcan (GTEx)")$gene_name

# Number of genes shared with those implicated above in single-tissue analysis 
(names(gene.tab) %in% txn.genes) %>% sum(.)
names(gene.tab)[(names(gene.tab) %in% txn.genes)] # 10 

# Number of genes NOT shared with those implicated above in single-tissue analysis 
txn.genes[!(txn.genes %in% names(gene.tab))] %>% length(.) # 19 

txn.uniq <- txn.genes[!(txn.genes %in% names(gene.tab))]

# Number of these unique genes that have been previously reported in GRASP/NHGRI-EBI catalogue 
filter(gws.df,tissue=="TissueXcan (GTEx)",gene_name%in%txn.uniq)$gene.type %>% table(.)
#novel.locus    reported  t2d.region 
#         10           6           3 
filter(gws.df,tissue=="TissueXcan (GTEx)",gene_name%in%txn.uniq,gene.type=="reported")$gene_name

```



### Bringing in Islet eQTLs for comparisons 

```{r}

# Islet is argueably most important tissue, did TWAS with islet eQTL models

# total number of significant associations/genes in islet analysis 
isl.genes <- filter(gws.df,tissue=="Islets")$gene_name
isl.genes %>% length(.) # 17 genes 


# number shared between Pancreas GWS genes 
panc.genes <- filter(gws.df,tissue=="Pancreas")$gene_name
isl.genes[isl.genes %in% panc.genes] %>% length(.)
shared.panc <- isl.genes[isl.genes %in% panc.genes]  
filter(gws.df,gene_name%in%shared.panc,tissue=="Islets")$gene.type

# number shared between TissueXcan GWS genes 
txn.genes <- filter(gws.df,tissue=="TissueXcan (GTEx)")$gene_name
isl.genes[isl.genes %in% txn.genes] %>% length(.)
shared.txn <- isl.genes[isl.genes %in% txn.genes]  
filter(gws.df,gene_name%in%shared.txn,tissue=="Islets")$gene.type

```





Conclusion 

```{r}

total.unique <- gws.df$gene_name %>% unique(.) %>% length(.)

filter(gws.df,tissue!="TissueXcan (GTEx)",!(gene_name%in%txn.genes))$gene_name %>% unique(.) %>% length(.)

sub.df <- filter(gws.df,tissue!="TissueXcan (GTEx)",!(gene_name%in%txn.genes)) %>% 
  dplyr::select(.,one_of("gene_name","gene.type"))
sub.df <- sub.df[!duplicated(sub.df$gene_name),]
sub.df$gene.type %>% table(.)
#novel.locus    reported  t2d.region 
 #        13           7           3 
```


### Profile all GWS genes implicated for novel discoveries 

```{r}

library("GenomicRanges")

region.df <- fread(dat.dir %&% "gwas_windows.txt")
len.vec <- region.df$loc.end - region.df$loc.start
len.vec %>%  sort(.)
len.vec %>% summary(.)

gws.df$gene_name %>% unique(.) %>% length(.)
sig.genes <-gws.df$gene_name %>% unique(.)  

# Breakdown of genes 
sub.df <- filter(gws.df,gene_name %in% sig.genes) %>% dplyr::select(.,one_of("gene_name","gene.type"))
sub.df <- sub.df[!duplicated(sub.df$gene_name),]
sub.df$gene.type %>% table(.)

# Evaluation of Novel genes 

novel.genes <- filter(sub.df,gene.type=="novel.locus")$gene_name

# how many of these novel genes are within 1Mb of gwas associations 

#reg.gr <- GRanges(seqnames=region.df$chrom,IRanges(start=region.df$loc.start,end=region.df$loc.end))
#expand.gr <- reg.gr + 1e6
#oneMb.vec <- map(novel.genes,function(g){
#  sub <- filter(assoc.df,gene_name==g)
#  if (g == "EGFEM1P"){
#    g.gr <- GRanges(seqnames=3,IRanges(start=167967729,end=168539123))
#  } else{
#    g.gr <- GRanges(seqnames=sub$chrom,IRanges(start=sub$gene.start,end=sub#$gene.end))
 # }
#  eval <- g.gr %over% expand.gr
#  if(eval==TRUE){
#    print(findOverlaps(expand.gr,g.gr))
#  }
#  return(eval)
#}) %>% as.logical(.) 

#novel.genes[oneMb.vec]

filter(gws.df,gene_name %in% novel.genes)$gene_name %>% table(.) %>% 
  sort(.) %>% rev(.)

filter(gws.df,gene_name %in% novel.genes)$tissue %>% table(.) %>% 
  sort(.) %>% rev(.)

filter(gws.df,gene_name %in% novel.genes,tissue=="TissueXcan (GTEx)")$gene_name
filter(gws.df,gene_name %in% novel.genes,tissue=="Islets")$gene_name
filter(gws.df,gene_name %in% novel.genes,tissue=="Pancreas")$gene_name
filter(gws.df,gene_name %in% novel.genes,tissue=="Liver")$gene_name
filter(gws.df,gene_name %in% novel.genes,tissue=="Muscle-Skeletal")$gene_name
filter(gws.df,gene_name %in% novel.genes,tissue=="Adipose-Subcutaneous")$gene_name

```

